### ISSUE TITLE
VAE encoding/decoding full example

### ISSUE THREAD

**Issue by staturecrane** [Wan-Video/Wan2.1#163]
Hello! Thank you so much for releasing this model and code.

I have an interest in using just the VAE you've developed for the model, and was hoping you could provide an example of the pre-processing you need to for encoding a video to a latent. I can see the post-process and decoding in the generation scripts, but it would be great to see how to utilize the encoder as well.

Thanks for any help you can provide!

**Comment by GooThinker** [comment]
To preprocess a video and encode it using the VAE (WanVAE) , we need to follow these steps:

1. Read the video: Use a library like opencv or torchvision to load the video frames.
2. Resize and normalize: Ensure the video frames are resized to the required dimensions and normalized to the range [-1, 1].
3. Reshape the video tensor: Convert the video into the shape [B, C, T, H, W] (batch size, channels, time frames, height, width).
4. Encode the video: Pass the preprocessed video tensor to the VAE's encode method. 

The following is a code implementation in Python, generated by Qwen 2.5:
`import cv2
import torch
import numpy as np
from .modules.model import WanModel

class VideoProcessor:
    def __init__(self, checkpoint_dir, config, device):
        # Initialize the VAE model
        self.vae = WanVAE(
            vae_pth=os.path.join(checkpoint_dir, config.vae_checkpoint),
            device=device
        )
        self.device = device

    def read_video(self, video_path, target_size=(128, 128), num_frames=None):
        """
        Reads a video file and preprocesses it for encoding.
        
        Args:
            video_path (str): Path to the video file.
            target_size (tuple): Target frame size (height, width).
            num_frames (int): Number of frames to extract (optional).
        
        Returns:
            torch.Tensor: Preprocessed video tensor with shape [1, C, T, H, W].
        """
        # Open the video file
        cap = cv2.VideoCapture(video_path)
        frames = []
        
        while True:
            ret, frame = cap.read()
            if not ret:
                break
            
            # Resize the frame
            frame = cv2.resize(frame, target_size)
            
            # Convert BGR to RGB
            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            
            # Normalize to [0, 1]
            frame = frame / 255.0
            
            # Normalize to [-1, 1]
            frame = (frame - 0.5) / 0.5
            
            frames.append(frame)
            
            # Stop if we have enough frames
            if num_frames is not None and len(frames) >= num_frames:
                break
        
        cap.release()
        
        # Stack frames into a tensor
        video_tensor = torch.tensor(np.stack(frames), dtype=torch.float32)  # Shape: [T, H, W, C]
        video_tensor = video_tensor.permute(3, 0, 1, 2)  # Shape: [C, T, H, W]
        video_tensor = video_tensor.unsqueeze(0)  # Add batch dimension: [1, C, T, H, W]
        
        return video_tensor.to(self.device)

    def encode_video(self, video_tensor):
        """
        Encodes a preprocessed video tensor using the VAE.
        
        Args:
            video_tensor (torch.Tensor): Preprocessed video tensor with shape [1, C, T, H, W].
        
        Returns:
            torch.Tensor: Encoded latent representation.
        """
        with torch.no_grad():
            latent = self.vae.encode(video_tensor)
        return latent


# Example usage
if __name__ == "__main__":
    # Configuration
    checkpoint_dir = "path/to/checkpoints"
    config = {
        "vae_checkpoint": "vae.pth"  # Replace with your VAE checkpoint filename
    }
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    
    # Initialize the processor
    processor = VideoProcessor(checkpoint_dir, config, device)
    
    # Read and preprocess the video
    video_path = "path/to/video.mp4"
    video_tensor = processor.read_video(video_path, target_size=(128, 128), num_frames=16)
    
    # Encode the video
    latent = processor.encode_video(video_tensor)
    print(f"Latent shape: {latent.shape}")
`

**Comment by Abhinay1997** [comment]
In case anyone feels opencv is slow, try decord.
